lib/client/db/postgres/mcp/mcp.go

```diff
@@ -0,0 +1,253 @@
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package mcp
+
+import (
+	"context"
+	"encoding/json"
+	"log/slog"
+	"net"
+	"time"
+
+	"github.com/gravitational/trace"
+	"github.com/jackc/pgx/v5"
+	"github.com/jackc/pgx/v5/pgxpool"
+	"github.com/mark3labs/mcp-go/mcp"
+
+	dbmcp "github.com/gravitational/teleport/lib/client/db/mcp"
+	"github.com/gravitational/teleport/lib/defaults"
+)
+
+// queryTool is the run query MCP tool definition.
+var queryTool = mcp.NewTool(dbmcp.ToolName(defaults.ProtocolPostgres, "query"),
+	mcp.WithDescription("Execute SQL query against PostgreSQL database connected using Teleport"),
+	mcp.WithString(queryToolDatabaseParam,
+		mcp.Required(),
+		mcp.Description("Teleport database resource URI where the query will be executed"),
+	),
+	mcp.WithString(queryToolQueryParam,
+		mcp.Required(),
+		mcp.Description("PostgresSQL SQL query to execute"),
+	),
+)
+
+type database struct {
+	*dbmcp.Database
+	pool *pgxpool.Pool
+}
+
+// Server PostgreSQL MCP server.
+type Server struct {
+	logger    *slog.Logger
+	databases map[string]*database
+}
+
+// NewServer initializes a PostgreSQL MCP server.
+func NewServer(ctx context.Context, cfg *dbmcp.NewServerConfig) (dbmcp.Server, error) {
+	s := &Server{logger: cfg.Logger, databases: make(map[string]*database)}
+
+	for _, db := range cfg.Databases {
+		if db.DatabaseUser == "" || db.DatabaseName == "" {
+			return nil, trace.BadParameter("must specify the username or database name used to connect to the database")
+		}
+
+		connCfg, err := buildConnConfig(db)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		pool, err := pgxpool.NewWithConfig(ctx, connCfg)
+		if err != nil {
+			return nil, trace.BadParameter("failed to parse database %q connection config: %s", db.DB.GetName(), err)
+		}
+
+		s.databases[db.ResourceURI()] = &database{
+			Database: db,
+			pool:     pool,
+		}
+	}
+
+	cfg.RootServer.AddTool(queryTool, s.RunQuery)
+	return s, nil
+}
+
+// Stop implements dbmcp.Server.
+func (s *Server) Close(context.Context) error {
+	for _, db := range s.databases {
+		db.pool.Close()
+	}
+
+	return nil
+}
+
+// RunQueryResult is the run query tool result.
+type RunQueryResult struct {
+	// Data contains the data returned from the query. It can be empty in case
+	// the query doesn't return any data.
+	Data []map[string]any `json:"data"`
+	// RowsCount number of rows affected by the query or returned as data.
+	RowsCount int `json:"rowsCount"`
+	// ErrorMessage if the query wasn't successful, this field contains the
+	// error message.
+	ErrorMessage string `json:"error,omitempty"`
+}
+
+// RunQuery tool function used to execute queries on databases.
+func (s *Server) RunQuery(ctx context.Context, request mcp.CallToolRequest) (*mcp.CallToolResult, error) {
+	uri := request.Params.Arguments[queryToolDatabaseParam].(string)
+	sql := request.Params.Arguments[queryToolQueryParam].(string)
```

Comment:
Nit: Use safe type assertion?

Reply:
Changed to use the newly added `RequireXXX` function that retries the request argument and asserts its type.